src/Http/Http.Abstractions/src/HttpRequest.cs

```diff
@@ -98,6 +98,24 @@ public abstract class HttpRequest
         /// <returns>The Content-Type header.</returns>
         public abstract string ContentType { get; set; }
 
+        /// <summary>
+        /// Indicates if the request may have a body.
+        /// </summary>
+        /// <remarks>
+        /// This returns true when:
+        /// - It's an HTTP/1.x request with a non-zero Content-Length or a 'Transfer-Encoding: chunked' header.
+        /// - It's an HTTP/2 request that did not set the END_STREAM flag on the initial headers frame.
+        /// The final request body length may still be zero for the chunked or HTTP/2 scenarios.
+        ///
+        /// This returns false when:
+        /// - It's an HTTP/1.x request with no Content-Length or 'Transfer-Encoding: chunked' header, or the Content-Length is 0.
+        /// - It's an HTTP/2 request that set END_STREAM on the initial headers frame.
+        /// When false, the request body should never return data.
+        ///
+        /// This returns null if the underlying server has not implemented the feature.
+        /// </remarks>
+        public virtual bool? CanHaveBody { get => null; }
```

Comment:
Is a property on HttpRequest necessary? This feels like a power-user feature that very few people would care about. Adding an extra property on request would be overkill.

If you care about this then getting the feature directly from the feature collection and using the interface is fine.

Reply:
I'd originally made it an extension method, but Fowler asked for a property.

We need to stop thinking about features as power user scenarios but rather as an implementation detail of HttpContext. You should be able to interact with all features via the HttpContext directly or via extensions. You should only interact with the feature interfaces directly if you're implementing or replacing them.

Reply:
> You should be able to interact with all features via the HttpContext directly or via extensions. 

I'm not sure I agree with that. There's a lot of features this doesn't apply to including those related to all sorts of TLS info, per-request rate limit configuration, subscribing to connection heartbeats, etc...